README.rst
khmer
=====

Welcome to khmer: k-mer counting, filtering and graph traversal FTW!

.. image:: https://badge.fury.io/py/khmer.png
    :target: http://badge.fury.io/py/khmer
.. image:: https://pypip.in/d/khmer/badge.png
    :target: https://crate.io/packages/khmer
.. image:: https://pypip.in/license/khmer/badge.png
    :target: https://pypi.python.org/pypi/khmer/
    :alt: License
.. image:: http://ci.ged.msu.edu/job/khmer/badge/icon
    :target: http://ci.ged.msu.edu/job/khmer/
.. image:: https://landscape.io/github/ged-lab/khmer/master/landscape.png
   :target: https://landscape.io/github/ged-lab/khmer/master
   :alt: Python Code Health
.. image:: https://scan.coverity.com/projects/621/badge.svg
   :target: https://scan.coverity.com/projects/621
   :alt: Coverity Scan Build Status

The official repository is at

https://github.com/ged-lab/khmer

and you can read the docs online here:

http://khmer.readthedocs.org/

There are two mailing lists dedicated to khmer, an announcements-only
list and a discussion list. To search their archives and sign-up for
them, please visit the following URLs:

-  Discussion: http://lists.idyll.org/listinfo/khmer

-  Announcements: http://lists.idyll.org/listinfo/khmer-announce

You can contact the current maintainers at khmer-project@idyll.org.

IMPORTANT NOTE: CITE US!
------------------------

khmer is *research software*, so you should cite us when you use it
in scientific publications!  Please see the CITATION file for
citation information.

INSTALL INSTRUCTIONS:
---------------------

khmer requires Python 2.7+. Linux users will need the python development
libraries and gcc. OS X users may need XCode installed. For more
details see `doc/install.txt <https://khmer.readthedocs.org/en/latest/install.html>`_

In short:

``pip install khmer`` to download, build, and install the latest stable
version.

The use of a virtualenv is recommended, see
http://www.virtualenv.org/en/latest/#installation

khmer is under the BSD license; see doc/LICENSE.txt. Distribution,
modification and redistribution, incorporation into other software, and
pretty much everything else is allowed.

khmer project contributors also edit documents collaboratively at `our wiki 
<https://github.com/ged-lab/khmer/wiki/_pages>`__.

MRC 2014-02-16

.. vim: set filetype=rst



ChangeLog
2014-04-01  Titus Brown  <t@Titus-MacBook-Air-2.local>

    * CITATION,khmer/khmer_args.py: Updated khmer software citation for
    release.

2014-03-31  Titus Brown  <t@Titus-MacBook-Air-2.local>

    * scripts/normalize-by-median.py: Fixed unbound variable bug introduced in
    20a433c2.

    * khmer/file.py: Fixed incorrect use of __file__ dirname instead of
    os.getcwd(); also fixed bug where statvfs would choke on an empty
    dirname resulting from input files being in the cwd.

2014-03-31  Michael R. Crusoe  <mcrusoe@msu.edu

    * versioneer.py,ez_setup.py: updated to version 0.10 and 3.4.1
    respectively.
    * docs/release.txt,khmer/_version.py,MANIFEST.in: update ancillary
    versioneer files

2014-03-31  Titus Brown  <t@Titus-MacBook-Air-2.local>

    * scripts/*.py,khmer/khmer_args.py: added 'info' function to khmer_args,
    and added citation information to each script.
    * CITATION: added basic citation information for khmer functionality.

2013-03-31  Michael R. Crusoe  <mcrusoe@msu.edu>

    * docs/scripts.txt,scripts/*.py,khmer/*.py: overhaul the documentation of
    the scripts. Uses sphinxcontrib.autoprogram to leverage the existing
    argparse objects. Moved the documentation into each script + misc cleanups.
    All scripts support the --version option. Migrated the last scripts to use
    khmer_args
    * docs/blog-posts.txt: removed outdated reference to filter-exact.py; its
    replacement filter-abund.py is better documented in the eel-pond protocol
    * figuregen/,novelty/,plots/,templatem/,scripts/do-partition.sh: removed
    outdated code not part of core project

2013-03-30  Michael R. Crusoe  <mcrusoe@msu.edu>

    * setup.py: monkeypatched distutils.Distribution.reinitialize_command() so
    that it matches the behavior of Distribution.get_command_obj(). This fixes
    issues with 'pip install -e' and './setup.py nosetests' not respecting the
    setup.cfg configuration directives for the build_ext command. Also
    enhanced our build_ext command to respect the dry_run mode.

    * .ycm_extra_conf.py: Update our custom YouCompleteMe configuration to
    query the package configuration for the proper compilation flags.

2014-03-28  Michael R. Crusoe  <mcrusoe@msu.edu>

    * Makefile,setup.py: demote nose & sphinx to extra dependencies.
    Auto-install Python developer tools as needed.

2013-03-27  Michael R. Crusoe  <mcrusoe@msu.edu>

    * The system zlib and bzip2 libraries are now used instead of the bundled
    versions if specified in setup.cfg or the command line.

2014-03-25  Michael R. Crusoe  <mcrusoe@msu.edu>

    * Makefile: update cppcheck command to match new version of Jenkins
    plugin. Now ignores the lib/test*.cc files.

2013-03-20  Michael R. Crusoe  <mcrusoe@msu.edu>

    * lib/storage.hh,khmer/_khmermodule.cc,lib/{readtable,read_parsers}.hh:
    remove unused storage.hh

2014-03-19  Qingpeng Zhang  <qingpeng@msu.edu>

    * hashbits.cc: fix a bug of 'Division or modulo by zero' described in #182
    * test_scripts.py: add test code for count-overlap.py
    * count-overlap.py: (fix a bug because of a typo and hashsize was replaced
    by min_hashsize)
    * count-overlap.py: needs hashbits table generated by load-graph.py. 
    This information is added to the "usage:" line.
    * count-overlap.py: fix minor PyLint issues

2014-03-19  Michael R. Crusoe  <mcrusoe@msu.edu>

    * Update bundled zlib version to 1.2.8 from 1.2.3. Changes of note:
    "Wholesale replacement of gz* functions with faster versions"
    "Added LFS (Large File Summit) support for 64-bit file offsets"
    "Fix serious but very rare decompression bug"

2014-03-19  Michael R. Crusoe <mcrusoe@msu.edu>

    * lib/counting.hh: include hashtable.hh
    * lib/{counting,aligner,hashbits,hashtable,labelhash,node,subset}.{cc,hh},
    kmer.cc,khmer/_khmermodule.cc: removed downcast, replaced non-functional
    asserts() with exception throws.
    * khmer/_khmermodule.cc: fixed parsing of PyLists
    * setup.py: force 64bit only builds on OS X.

2014-03-19  Titus Brown  <t@idyll.org>

    * Makefile: update documentation on targets at top; clean autopep8 output.
    * test_counting_single.py: fixed pep8 violations in spacing
    * test_scripts.py: eliminate popenscript in favor of proper SystemExit
	handling in runscript; fix pep8 violations.

2014-03-19  Michael R. Crusoe <mcrusoe@msu.edu> and Luiz Irber
<luiz.irber@gmail.com>

    * lib/ktable.{cc,hh},khmer/{__init__.py},{_khmermodule.cc}, tests/
    test_{counting_{hash,single},ktable}.py: remove the unused KTable object
    * doc/{index,ktable}.txt: remove references to KTable
    * lib/{ktable.{hh,cc} → kmer_hash.{hh,cc}}: rename remaining ktable files
    to kmer_hash
    * lib/{hashtable,kmer}.hh: replace ktable headers with kmer_hash

2014-03-17  Ram RS  <ramrs@nyu.edu>

    * extract-partitions.py: pylint warnings addressed
    * test_scripts.py: tests added to cover extract-partitions completely

2014-03-16  Michael R. Crusoe <mcrusoe@msu.edu>

    * lib/read_parsers.cc: fix for Coverity CID 1054789: Unititialized scalar
    field II: fill_id is never zeroed out.

2014-03-16  Ram RS  <ramrs@nyu.edu>

    * Project email in copyright headers updated

2014-03-14  Michael R. Crusoe <mcrusoe@msu.edu>

    * khmer/_khmermodule.cc, lib/{khmer.hh, hashtable.{cc,hh}},
    tests/test_{hashbits,hashbits_obj,labelhash}.py: don't implicitly downcast
    tagset_size(). Changes fileformat version for saved tagsets.

2014-03-13  Ram RS  <ramrs@nyu.edu>

    * added: khmer/file.py - script to check disk space, check input file
    status and check space before hashtable writing
    * modified: scripts/*.py - all scripts now use khmer.file for above-mentioned
    functionality.
    * modified: scripts/*.py - pylint violations addressed in all scripts
    under scripts/

2014-03-13  Ram RS  <ramrs@nyu.edu>

    * Bug fix: tests.test_normalize_by_median_no_bigcount() now runs within
    temp directory


sandbox/test_traverse.py
import khmer

def rc(kmer):
    ret = ''
    for i in range(len(kmer)-1,-1,-1):
        base = kmer[i]
        if base == 'A':
            ret += 'T'
        elif base == 'C':
            ret += 'G'
        elif base == 'G':
            ret += 'C'
        else:
            ret += 'A'
    return ret
    
K = 20
seq = 'ATCGTCGATCTACGACTACGACTCGATCGATCGACTCGATCGATC'

ind = 2

start = seq[ind:ind+K]
assert len(start) == K

left = seq[ind-1:ind-1+K]
right = seq[ind+1:ind+1+K]

assert len(left) == K
assert len(right) == K

lb = khmer.LabelHash(K, 1e9, 4)
lb.consume(seq)

tr_left = lb.traverse_from_kmer(start, 'l')
tr_right = lb.traverse_from_kmer(start, 'r')

print 'sequence', seq
print 'start', start
print '*' * 40

print 'expected L:', left, rc(left)
print 'actual L:', tr_left, rc(tr_left[0])
print '*' * 40

print 'expected R:', right, rc(right)
print khmer.forward_hash(right, K)
print khmer.forward_hash(rc(right), K)
print lb.get(right)
print lb.get(rc(right))

print seq.find(right)

print '=' * 40
print 'actual R:', tr_right[0], rc(tr_right[0])
print khmer.forward_hash(tr_right[0], K)
print khmer.forward_hash(rc(tr_right[0]), K)
print lb.get(tr_right[0])
print lb.get(rc(tr_right[0]))

print seq.find(tr_right[0])
print seq.find(rc(tr_right[0]))


lib/labelhash.cc
//
// This file is part of khmer, http://github.com/ged-lab/khmer/, and is
// Copyright (C) Michigan State University, 2009-2013. It is licensed under
// the three-clause BSD license; see doc/LICENSE.txt. Contact: ctb@msu.edu
//

#include "labelhash.hh"

#define LABEL_DBG 0
#define printdbg(m) if(LABEL_DBG) std::cout << #m << std::endl;

using namespace khmer;
using namespace khmer:: read_parsers;

/*
 * @camillescott
 * Might be time for a refactor: could do a general consume_fasta
 * function which accepts a consume_sequence function pointer as a parameter
 */

void
LabelHash::consume_fasta_and_tag_with_labels(
  std:: string const  &filename,
  unsigned int	      &total_reads, unsigned long long	&n_consumed,
  CallbackFn	      callback,	    void *		callback_data
)
{
  khmer:: Config    &the_config	  = khmer:: get_active_config( );

  // Note: Always assume only 1 thread if invoked this way.
  IParser *	  parser = 
  IParser::get_parser(
    filename, 1, the_config.get_reads_input_buffer_size( ),
    the_config.get_reads_parser_trace_level( )
  );


  consume_fasta_and_tag_with_labels(
    parser,
    total_reads, n_consumed,
    callback, callback_data
  );

  delete parser;
}

void
LabelHash::consume_fasta_and_tag_with_labels(
    read_parsers:: IParser *  parser,
    unsigned int		    &total_reads,   unsigned long long	&n_consumed,
    CallbackFn		    callback,	    void *		callback_data
  )
  {
    Hasher		  &hasher		= 
    _get_hasher( parser->uuid( ) );
    unsigned int		  total_reads_LOCAL	= 0;
  #if (0) // Note: Used with callback - currently disabled.
    unsigned long long int  n_consumed_LOCAL	= 0;
  #endif
    Read			  read;

    // TODO? Delete the following assignments.
    total_reads = 0;
    n_consumed = 0;
    
    hasher.trace_logger(
      TraceLogger:: TLVL_DEBUG2,
      "Starting trace of 'consume_fasta_and_tag_with_labels'....\n"
    );
    
    Label _tag_label = 0;

    Label * the_label;
    // Iterate through the reads and consume their k-mers.
    while (!parser->is_complete( ))
    {
      unsigned long long this_n_consumed   = 0;

      read = parser->get_next_read( );

      if (check_and_normalize_read( read.sequence ))
      {
        // TODO: make threadsafe!
        the_label = check_and_allocate_label(_tag_label);
        consume_sequence_and_tag_with_labels( read.sequence,
					      this_n_consumed,
					      *the_label );
	    _tag_label++;

  #ifdef WITH_INTERNAL_METRICS
        hasher.pmetrics.start_timers( );
  #endif
  #if (0) // Note: Used with callback - currently disabled.
        n_consumed_LOCAL  = __sync_add_and_fetch( &n_consumed, this_n_consumed );
  #else
        __sync_add_and_fetch( &n_consumed, this_n_consumed );
  #endif
        total_reads_LOCAL = __sync_add_and_fetch( &total_reads, 1 );
  #ifdef WITH_INTERNAL_METRICS
        hasher.pmetrics.stop_timers( );
        hasher.pmetrics.accumulate_timer_deltas(
	  (uint32_t)HashTablePerformanceMetrics:: MKEY_TIME_UPDATE_TALLIES
        );
  #endif
      }

      if (0 == (total_reads_LOCAL % 10000))
        hasher.trace_logger(
	  TraceLogger:: TLVL_DEBUG3,
	  "Total number of reads processed: %llu\n",
	  (unsigned long long int)total_reads_LOCAL
        );

      // TODO: Figure out alternative to callback into Python VM
      //       Cannot use in multi-threaded operation.
  #if (0)
        // run callback, if specified
        if (total_reads_TL % CALLBACK_PERIOD == 0 && callback) {
	  std::cout << "n tags: " << all_tags.size() << "\n";
	  try {
	    callback("consume_fasta_and_tag_with_labels", callback_data, total_reads_TL,
		     n_consumed);
	  } catch (...) {
	    delete parser;
	    throw;
	  }
        }
  #endif // 0

    } // while reads left for parser

  }

void LabelHash::consume_partitioned_fasta_and_tag_with_labels(const std::string &filename,
					  unsigned int &total_reads,
					  unsigned long long &n_consumed,
					  CallbackFn callback,
					  void * callback_data)
{
  total_reads = 0;
  n_consumed = 0;

  IParser* parser = IParser::get_parser(filename.c_str());
  Read read;

  std::string seq = "";

  // reset the master subset partition
  //delete partition;
  //partition = new SubsetPartition(this);

  //
  // iterate through the FASTA file & consume the reads.
  //
  Label * c;
  PartitionID p;
  while(!parser->is_complete())  {
    read = parser->get_next_read();
    seq = read.sequence;

    if (check_and_normalize_read(seq)) {
      // First, figure out what the partition is (if non-zero), and save that.
      printdbg(parsing partition id)
      p = _parse_partition_id(read.name);
      printdbg(checking label and allocating if necessary)
      c = check_and_allocate_label(p);
      printdbg(consuming sequence and tagging)
      consume_sequence_and_tag_with_labels( seq,
					      n_consumed,
					      *c );
      printdbg(back in consume_partitioned)
    }
	       
    // reset the sequence info, increment read number
    total_reads++;

    // run callback, if specified
    if (total_reads % CALLBACK_PERIOD == 0 && callback) {
      try {
        callback("consume_partitioned_fasta_and_tag_with_labels", callback_data, 
        total_reads, n_consumed);
      } catch (...) {
	delete parser;
        throw;
      }
    }
  }
  printdbg(done with while loop in consume_partitioned)

  // @cswelcher TODO: check that deallocate LabelPtrMap is correct
  delete parser;
  printdbg(deleted parser and exiting)
}

// @cswelcher: double-check -- is it valid to pull the address from a reference?
void LabelHash::link_tag_and_label(HashIntoType& kmer, Label& kmer_label) {
    printdbg(linking tag and label)
    tag_labels.insert(TagLabelPtrPair(kmer, &kmer_label));
    label_tag_ptrs.insert(LabelTagPtrPair(kmer_label, &kmer));
    printdbg(done linking tag and label)
}

void LabelHash::consume_sequence_and_tag_with_labels(const std::string& seq,
					unsigned long long& n_consumed,
					Label& current_label,
					SeenSet * found_tags)
  {

    printdbg(inside low-level labelhash consume sequence function)

    bool is_new_kmer;
    bool kmer_tagged;

    KMerIterator kmers(seq.c_str(), _ksize);
    HashIntoType kmer;

    unsigned int since = _tag_density / 2 + 1;
    
    printdbg(entering while loop)
    while(!kmers.done()) {
      kmer = kmers.next();

      if ((is_new_kmer = test_and_set_bits( kmer ))) {
        ++n_consumed;
        printdbg(test_and_set_bits)
      }
  #if (1)
      if (is_new_kmer) {
        printdbg(new kmer...)
        ++since;
      } else {
        printdbg(entering tag spin lock)
        //ACQUIRE_ALL_TAGS_SPIN_LOCK
        kmer_tagged = set_contains(all_tags, kmer);
        //RELEASE_ALL_TAGS_SPIN_LOCK
        printdbg(released tag spin lock)
        if (kmer_tagged) {
	      since = 1;
	      printdbg(kmer already in all_tags)
	      // Labeling code
	      // TODO: MAKE THREADSAFE!
	      
	      if (!_cmap_contains_label(tag_labels, kmer, current_label)) {
            printdbg(tag was not labeled: adding to labels...)
	        //ACQUIRE_TAG_COLORS_SPIN_LOCK
	        link_tag_and_label(kmer, current_label);
	        //RELEASE_TAG_COLORS_SPIN_LOCK
            printdbg(released label spin lock)
	      }
	      if (found_tags) {
	        found_tags->insert(kmer);
	      }
        }  else {
            printdbg(inc since var)
            ++since;
        }
      }
  #else
      if (!is_new_kmer && set_contains(all_tags, kmer)) {
        since = 1;
        if (found_tags) { found_tags->insert(kmer); }
      } else {
        since++;
      }
  #endif
      //
      if (since >= _tag_density) {
        printdbg(exceeded tag density: drop a tag and label -- getting tag lock)
        //ACQUIRE_ALL_TAGS_SPIN_LOCK
        printdbg(in tag spin lock)
        all_tags.insert(kmer);
        //RELEASE_ALL_TAGS_SPIN_LOCK
        printdbg(released tag spin lock)
        
        // Labeling code
        // TODO: MAKE THREADSAFE!
        //ACQUIRE_TAG_COLORS_SPIN_LOCK
        link_tag_and_label(kmer, current_label);
        //RELEASE_TAG_COLORS_SPIN_LOCK
        
        if (found_tags) { found_tags->insert(kmer); }
        since = 1;
      }
    printdbg(moving to next iter)
    } // iteration over kmers
    printdbg(finished iteration: dropping last tag)
    if (since >= _tag_density/2 - 1) {
      //ACQUIRE_ALL_TAGS_SPIN_LOCK
      all_tags.insert(kmer);	// insert the last k-mer, too.
      //RELEASE_ALL_TAGS_SPIN_LOCK
      
      // Label code: TODO: MAKE THREADSAFE!
      link_tag_and_label(kmer, current_label);
      
      if (found_tags) { found_tags->insert(kmer); }
    }
  printdbg(done with low-level consume)
}

unsigned int LabelHash::sweep_label_neighborhood(const std::string& seq,
                                                  LabelPtrSet& found_labels,
                                                  unsigned int range,
                                                  bool break_on_stoptags,
                                                  bool stop_big_traversals) {

    SeenSet tagged_kmers;
    unsigned int num_traversed;
    num_traversed = partition->sweep_for_tags(seq, tagged_kmers, all_tags, 
                              range, break_on_stoptags, stop_big_traversals);
    traverse_labels_and_resolve(tagged_kmers, found_labels);
    //printf("range=%u ", range);
    if (range == 0) {
      assert(num_traversed == seq.length()-ksize()+1);
    }
    tagged_kmers.clear();
    return num_traversed;
}

LabelPtrSet LabelHash::get_tag_labels(const HashIntoType& tag) {
  LabelPtrSet labels;
  //unsigned int num_labels;
  _get_tag_labels(tag, tag_labels, labels);
  return labels;
}

TagPtrSet LabelHash::get_label_tags(const Label& label) {
  TagPtrSet tags;
  //unsigned int num_tags;
  _get_tags_from_label(label, label_tag_ptrs, tags);
  return tags;
}

void LabelHash::traverse_labels_and_resolve(const SeenSet& tagged_kmers,
                                              LabelPtrSet& found_labels) {
  
  SeenSet::const_iterator si;
  unsigned int num_labels = 0;
  for (si=tagged_kmers.begin(); si!=tagged_kmers.end(); ++si) {
    HashIntoType tag = *si;
    // get the labels associated with this tag
    num_labels = _get_tag_labels(tag, tag_labels, found_labels);
    if (num_labels > 1) {
      // reconcile labels
      // for now do nothing ha
    }
  }
}


void LabelHash::traverse_right(const HashIntoType& kmer_f,
                               const HashIntoType& kmer_r,
                               SeenSet& neighbors) {
    HashIntoType f, r;
    const unsigned int rc_left_shift = _ksize*2 - 2;
    //const HashIntoType bitmask = bitmask;
  
    f = next_f(kmer_f, 'A');
    r = next_r(kmer_r, 'A');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }

    f = next_f(kmer_f, 'C');
    r = next_r(kmer_r, 'C');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }

    f = next_f(kmer_f, 'T');
    r = next_r(kmer_r, 'T');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }

    f = next_f(kmer_f, 'G');
    r = next_r(kmer_r, 'G');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }
}

void LabelHash::traverse_left(const HashIntoType& kmer_f,
                              const HashIntoType& kmer_r,
                              SeenSet& neighbors) {
    HashIntoType f, r;
    const unsigned int rc_left_shift = _ksize*2 - 2;
    
    f = prev_f(kmer_f, 'A');
    r = prev_r(kmer_r, 'A');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }

    f = prev_f(kmer_f, 'C');
    r = prev_r(kmer_r, 'C');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }

    f = prev_f(kmer_f, 'T');
    r = prev_r(kmer_r, 'T');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }

    f = prev_f(kmer_f, 'G');
    r = prev_r(kmer_r, 'G');
    if (get_count(uniqify_rc(f,r))) {
      neighbors.insert(uniqify_rc(f,r));
    }
}
 


scripts/sandbox/fix_normalize.py
    n = -1
    total = 0
    discarded = 0

    hb = khmer.new_hashbits(ht.ksize(),1,1)

    for n, record in enumerate(screed.open(
        input_filename))

    hb.consume(record.sequence)
        #how can i get the counts of a batch?
        if get_ratio(record.sequence,K,hb) > .50
            discarded += 1
        else
            ht.consume(record.sequence)
            #load kmers of count > C into hashbits
            #(TODO add C & hb as args)
            #find all kmers of K count (in ht) above C, load into hb

        if n > 0 and n % 100000 == 0:
            print '... kept {kept} of {total} or {perc:2}%'.format(
                kept=total - discarded, total=total,
                perc=int(100. - discarded / float(total) * 100.))
            print '... in file', input_filename

            if report_fp:
                print>>report_fp, total, total - discarded, \
                    1. - (discarded / float(total))
                report_fp.flush()

                hb.HT_SIZE += random.randint(-50,50)               #added
                        total += batch_size
                        #-----------------------------    
                        hb = khmer.new_hashbits(ht.ksize(),1,1)
                        assert hb
                        if not hb:
                            raise AssertionError("hashbits is not so much fun right now")

                        for n, record in enumerate(screed.open(
                            input_filename))

                        hb.consume(record.sequence)
                        #how can i get the counts of a batch?
                        ###kmers = get_counts(batch)
                        #from there i need to count the number of 0's vs 1's (present,
                        #not_present)
                        if(present > not_present)
                        discarded += 1
                    else
                        ht.consume(batch)
                        #load kmers of count > C into hashbits
                        #(TODO add C & hb as args)
                        for 
                        ###kmers.consume(above_C)

                        HT_SIZE += random.randint(-50,50)               #added

        if n> 100000 and n % 10000 == 0


scripts/sandbox/broken_normalize.py
#! /usr/bin/env python
#
# This file is part of khmer, http://github.com/ged-lab/khmer/, and is
# Copyright (C) Michigan State University, 2009-2013. It is licensed under
# the three-clause BSD license; see doc/LICENSE.txt. Contact: ctb@msu.edu
#
"""
Eliminate reads with median k-mer abundance higher than
DESIRED_COVERAGE.  Output sequences will be placed in 'infile.keep'.

% python scripts/normalize-by-median.py [ -C <cutoff> ] <data1> <data2> ...

Use '-h' for parameter help.
"""

import sys
import screed
import os
import random
import khmer
from itertools import izip
from khmer.counting_args import build_construct_args, DEFAULT_MIN_HASHSIZE
import argparse

DEFAULT_DESIRED_COVERAGE = 10

# Iterate a collection in arbitrary batches
# from: http://stackoverflow.com/questions/4628290/pairs-from-single-list


def batchwise(t, size):
    it = iter(t)
    return izip(*[it] * size)

# Returns true if the pair of records are properly pairs


def validpair(r0, r1):
    return r0.name[-1] == "1" and \
        r1.name[-1] == "2" and \
        r0.name[0:-1] == r1.name[0:-1]

def kmers(seq,K):
    for i in xrange(len(seq)-K+1):
        yield seq[i:i+1]

def get_ratio(seq,K,hb):
    sum = 0.0
        for kmer in kmers(S,25):
            sum += hb.get(kmer)
        return sum/float(len(seq-K+1))

def normalize_by_median(input_filename, outfp, ht, args, report_fp=None):


    DESIRED_COVERAGE = args.cutoff
    K = ht.ksize()

    # In paired mode we read two records at a time
    batch_size = 1
    if args.paired:
        batch_size = 2

    hb = khmer.new_hashbits(ht.ksize(),1,1)
    assert hb
    if not hb:
        raise AssertionError("hashbits is not so much fun right now")

    n = -1
    total = 0
    discarded = 0

    for n, batch in enumerate(batchwise(screed.open(
            input_filename), batch_size)):

'''
        hb.consume(batch)                                       #initiate hb
        #how can i get the counts of a batch?
        ###kmers = get_counts(batch)
        #from there i need to count the number of 0's vs 1's (present,
        #not_present)
        if(present > not_present)
            discarded += 1
        else
            ht.consume(batch)
            #load kmers of count > C into hashbits
            #(TODO add C & hb as args)
            for 
            ###kmers.consume(above_C)
'''
        if n > 0 and n % 100000 == 0:
            print '... kept {kept} of {total} or {perc:2}%'.format(
                kept=total - discarded, total=total,
                perc=int(100. - discarded / float(total) * 100.))
            print '... in file', input_filename

            if report_fp:
                print>>report_fp, total, total - discarded, \
                    1. - (discarded / float(total))
                report_fp.flush()

            hb.HT_SIZE += random.randint(-50,50)               #added
        total += batch_size

        # If in paired mode, check that the reads are properly interleaved
        if args.paired:
            if not validpair(batch[0], batch[1]):
                raise IOError('Error: Improperly interleaved pairs \
                    {b0} {b1}'.format(b0=batch[0].name, b1=batch[1].name))

        # Emit the batch of reads if any read passes the filter
        # and all reads are longer than K
        passed_filter = False
        passed_length = True
        for record in batch:
            if len(record.sequence) < K:
                passed_length = False
                continue

            seq = record.sequence.replace('N', 'A')
            med, _, _ = ht.get_median_count(seq)

            if med < DESIRED_COVERAGE:
                ht.consume(seq)
                passed_filter = True

            if report_fp:                                               #added
                print>>report_fp, total, total - discarded, \           #added  
                    1. - (discarded / float(total))                     #added
                report_fp.flush()                                       #added 

        # Emit records if any passed
        if passed_length and passed_filter:
            for record in batch:
                if hasattr(record, 'accuracy'):
                    outfp.write(
                        '@{name}\n{seq}\n'
                        '+\n{acc}\n'.format(name=record.name,
                                            seq=record.sequence,
                                            acc=record.accuracy))
                else:
                    outfp.write(
                        '>{name}\n{seq}\n'.format(name=record.name,
                                                  seq=record.sequence))
        else:
            discarded += batch_size

    return total, discarded


def handle_error(error, output_name, input_name, ht):
    print >>sys.stderr, '** ERROR:', error
    print >>sys.stderr, '** Failed on {name}: '.format(name=input_name)
    hashname = os.path.basename(input_name) + '.ht.failed'
    print >>sys.stderr, '** ...dumping hashtable to {ht}'.format(ht=hashname)
    ht.save(hashname)
    try:
        os.remove(output_name)
    except:
        print >>sys.stderr, '** ERROR: problem removing erroneous .keep file'


def main():
    parser = build_construct_args()
    parser.add_argument('-C', '--cutoff', type=int, dest='cutoff',
                        default=DEFAULT_DESIRED_COVERAGE)
    parser.add_argument('-p', '--paired', action='store_true')
    parser.add_argument('-s', '--savehash', dest='savehash', default='')
    parser.add_argument('-l', '--loadhash', dest='loadhash',
                        default='')
    parser.add_argument('-R', '--report-to-file', dest='report_file',
                        type=argparse.FileType('w'))
    parser.add_argument('-f', '--force-processing', dest='force',
                        help='continue on next file if read errors are \
                         encountered', action='store_true')
    parser.add_argument('-d', '--dump-frequency', dest='dump_frequency',
                        type=int, help='dump hashtable every d files',
                        default=-1)
    parser.add_argument('input_filenames', nargs='+')

    args = parser.parse_args()

    if not args.quiet:
        if args.min_hashsize == DEFAULT_MIN_HASHSIZE and not args.loadhash:
            print >>sys.stderr, \
                "** WARNING: hashsize is default!  " \
                "You absodefly want to increase this!\n** " \
                "Please read the docs!"

        print >>sys.stderr, '\nPARAMETERS:'
        print >>sys.stderr, \
            ' - kmer size =    {ksize:d} \t\t(-k)'.format(ksize=args.ksize)
        print >>sys.stderr, \
            ' - n hashes =     {nhash:d} \t\t(-N)'.format(nhash=args.n_hashes)
        print >>sys.stderr, \
            ' - min hashsize = {mh:-5.2g} \t(-x)'.format(mh=args.min_hashsize)
        print >>sys.stderr, ' - paired = {pr} \t\t(-p)'.format(pr=args.paired)
        print >>sys.stderr, ''
        print >>sys.stderr, \
            'Estimated memory usage is {prod:.2g} bytes \
            (n_hashes x min_hashsize)'.format(prod=args.n_hashes
                                              * args.min_hashsize)
        print >>sys.stderr, '-' * 8

    K = args.ksize
    HT_SIZE = args.min_hashsize
    N_HT = args.n_hashes

    report_fp = args.report_file
    filenames = args.input_filenames
    force = args.force
    dump_frequency = args.dump_frequency

    # list to save error files along with throwing exceptions
    if force is True:
        corrupt_files = []

    if args.loadhash:
        print 'loading hashtable from', args.loadhash
        ht = khmer.load_counting_hash(args.loadhash)
    else:
        print 'making hashtable'
        ht = khmer.new_counting_hash(K, HT_SIZE, N_HT)

    total = 0
    discarded = 0

    for n, input_filename in enumerate(filenames):
        output_name = os.path.basename(input_filename) + '.keep'
        outfp = open(output_name, 'w')

        total_acc = 0
        discarded_acc = 0

        try:
            total_acc, discarded_acc = normalize_by_median(input_filename,
                                                           outfp, ht, args,
                                                           report_fp)
        except IOError as e:
            handle_error(e, output_name, input_filename, ht)
            if not force:
                print >>sys.stderr, '** Exiting!'
                sys.exit(-1)
            else:
                print >>sys.stderr, '*** Skipping error file, moving on...'
                corrupt_files.append(input_filename)
                pass
        else:
            if total_acc == 0 and discarded_acc == 0:
                print 'SKIPPED empty file', input_filename
            else:
                total += total_acc
                discarded += discarded_acc
                print 'DONE with {inp}; kept {kept} of {total} or {perc:2}%'\
                    .format(inp=input_filename,
                            kept=total - discarded, total=total,
                            perc=int(100. - discarded / float(total) * 100.))
                print 'output in', output_name

        if dump_frequency > 0 and n > 0 and n % dump_frequency == 0:
            print 'Backup: Saving hashfile through', input_filename
            if args.savehash:
                hashname = args.savehash
                print '...saving to', hashname
            else:
                hashname = 'backup.ht'
                print 'Nothing given for savehash, saving to', hashname
            ht.save(hashname)

    if args.savehash:
        print 'Saving hashfile through', input_filename
        print '...saving to', args.savehash
        ht.save(args.savehash)

    # Change 0.2 only if you really grok it.  HINT: You don't.
    fp_rate = khmer.calc_expected_collisions(ht)
    print 'fp rate estimated to be {fpr:1.3f}'.format(fpr=fp_rate)

    if force and len(corrupt_files) > 0:
        print >>sys.stderr, "** WARNING: Finished with errors!"
        print >>sys.stderr, "** IOErrors occurred in the following files:"
        print >>sys.stderr, "\t", " ".join(corrupt_files)

    if fp_rate > 0.20:
        print >>sys.stderr, "**"
        print >>sys.stderr, "** ERROR: the counting hash is too small for"
        print >>sys.stderr, "** this data set.  Increase hashsize/num ht."
        print >>sys.stderr, "**"
        print >>sys.stderr, "** Do not use these results!!"
        sys.exit(-1)

if __name__ == '__main__':
    main()

# vim: set ft=python ts=4 sts=4 sw=4 et tw=79:
